storage/cloud-client/encryption.py

```diff
@@ -0,0 +1,161 @@
+        source_file_name, encryption_key=encryption_key)
+
+    print('File {} uploaded to {}.'.format(
+        source_file_name,
+        destination_blob_name))
+
+
+def download_encrypted_blob(bucket_name, source_blob_name,
+                            destination_file_name, base64_encryption_key):
+    """Downloads a previously-encrypted blob from Google Cloud Storage.
+
+    The encryption key provided must be the same key provided when uploading
+    the blob.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(bucket_name)
+    blob = bucket.blob(source_blob_name)
+
+    # Encryption key must be an AES256 key represented as a bytestring with
+    # 32 bytes. Since it's passed in as a base64 encoded string, it needs
+    # to be decoded.
+    encryption_key = base64.b64decode(base64_encryption_key)
+
+    blob.download_to_filename(
+        destination_file_name, encryption_key=encryption_key)
+
+    print('Blob {} downloaded to {}.'.format(
+        source_blob_name,
+        destination_file_name))
+
+
+def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
+                          base64_new_encryption_key):
+    """Performs a key rotation by re-writing an encrypted blob with a new
+    encryption key."""
+    raise NotImplementedError(
+        'This is currently not available using the Cloud Client Library.')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    subparsers = parser.add_subparsers(dest='command')
+
+    subparsers.add_parser(
+        'generate-encryption-key', help=generate_encryption_key.__doc__)
+
+    upload_parser = subparsers.add_parser(
+        'upload', help=upload_encrypted_blob.__doc__)
+    upload_parser.add_argument(
+        'bucket_name', help='Your cloud storage bucket.')
+    upload_parser.add_argument('source_file_name')
+    upload_parser.add_argument('destination_blob_name')
+    upload_parser.add_argument('base64_encryption_key')
+
+    download_parser = subparsers.add_parser(
+        'download', help=download_encrypted_blob.__doc__)
+    download_parser.add_argument(
+        'bucket_name', help='Your cloud storage bucket.')
+    download_parser.add_argument('source_blob_name')
+    download_parser.add_argument('destination_file_name')
+    download_parser.add_argument('base64_encryption_key')
+
+    rotate_parser = subparsers.add_parser(
+        'rotate', help=rotate_encryption_key.__doc__)
+    rotate_parser.add_argument(
+        'bucket_name', help='Your cloud storage bucket.')
+    download_parser.add_argument('blob_name')
+    download_parser.add_argument('base64_encryption_key')
+    download_parser.add_argument('base64_new_encryption_key')
+
+    args = parser.parse_args()
+
+    if args.command == 'generate-encryption-key':
+        generate_encryption_key()
+    elif args.command == 'upload':
```

Comment:
Yeah, so I explicitly picked this pattern over that one.

I don't like our functions just taking args. It's less clear, it makes testing slightly harder, and it's just doesn't feel right. We've worked around this in the past by having a '_command' function, eg:

```
list_objects(bucket_name):
    ...

list_objects_command(args):
    list_objects(args.bucket_name)
```

but I don't like having to write those either. This is more direct and simple, but I'm open to arguments against it or for something else.